### PR TITLE
Added multiGPU support compatible with optnet

### PR DESCRIPTION
**Pull request by melgor**
### What does this PR do?

- Added multi-GPU support (nGPU option), based on https://github.com/szagoruyko/fb.resnet.torch/tree/optnet 
- refactor code. It was needed because of merging optnet with MultiGPU. Here I need to prepare the model before every epoch. So I need to make it possible.
- remove changing model to float (as I have trouble with it and I do not think that anyone will train model using CPU)
- break test of LFW (as here the float model was used and when using multiGPU, we need some more logic to be working with it. Additional, I will propose new method of testing the Verification accuracy using any data provided, but with same format. It will remove dependece of using Python code in training module)
- add option "cudnn_bench". It speedup learnig by ~3% but require additional memory

### Where should the reviewer start?

It should start with train.lua and model.lua.

### How should this PR be tested?
It should be tested by running learning of same model with current code and new one. The LFW testing should be turned off

### Any background context you want to provide?
I mange to get 2x batch size compared to 1 GPU:)

### What are the relevant issues?
#106

### CODE REVIEW

File: training/model.lua [hunk 1]
```diff
@@ -4,47 +4,56 @@ require 'dpnn'
...
-   if opt.cudnn then
-      require 'cudnn'
-      cudnn.benchmark = false
-      cudnn.fastest = true
-      cudnn.verbose = false
-   end
-end
-
-paths.dofile('torch-TripletEmbedding/TripletEmbedding.lua')
+require 'cunn'
```
Comment (bamos) [inline training/model.lua]: Can you add back the CUDA guards? Travis CI tests a few iterations of training in CPU mode and also some users use CPU mode for fine-tuning.

File: training/train.lua [hunk 2]
```diff
@@ -107,25 +104,36 @@ function train()
...
+ 
+   local dpt
+   if torch.type(model) == 'nn.DataParallelTable' then
+      dpt   = model
+      model = model:get(1)        
+   end    
+   
+   local optnet_loaded, optnet = pcall(require,'optnet')
+   if optnet_loaded then
+    optnet.removeOptimization(model)
```
Comment (bamos) [inline training/train.lua]: Is the optimization ever added again? Since this isn't on a cloned model, will this remove the optnet improvements in the model being trained the first time this is called and never add them back?

File: training/util.lua [hunk 3]
```diff
@@ -70,3 +70,37 @@ function optimizeNet( model, inputSize )
...
+         end)
+      dpt.gradInput = nil
+
+      model = dpt:cuda()
+   end
+   return model
+end
+
+function MSRinit(model)
+   for k,v in pairs(model:findModules('nn.SpatialConvolution')) do
```
Comment (bamos) [inline training/util.lua]: Some modules use `nn.SpatialConvolutionMM` and `torch.findModules` won't initialize these:

```lua
require 'nn'
net = nn.Sequential()
net:add(nn.SpatialConvolutionMM(3, 16, 5, 5))
input = torch.rand(3,128,128)
out = net:forward(input)
net:findModules('nn.SpatialConvolution')
net:findModules('nn.SpatialConvolutionMM')
```

```lua
th> net:findModules('nn.SpatialConvolution')
{}
{}
                                                                      [0.0001s]
th> net:findModules('nn.SpatialConvolutionMM')
{
  1 :
    {
      dH : 1
      dW : 1
      nOutputPlane : 16
      output : DoubleTensor - size: 16x124x124
      gradInput : DoubleTensor - empty
      fgradInput : DoubleTensor - empty
      finput : DoubleTensor - size: 75x15376
      gradBias : DoubleTensor - size: 16
      padH : 0
      weight : DoubleTensor - size: 16x75
      bias : DoubleTensor - size: 16
      gradWeight : DoubleTensor - size: 16x75
      padW : 0
      nInputPlane : 3
      kW : 5
      kH : 5
    }
}
{
  1 :
    {
      gradInput : DoubleTensor - empty
      modules :
        {
          1 :
            {
              dH : 1
              dW : 1
              nOutputPlane : 16
              output : DoubleTensor - size: 16x124x124
              gradInput : DoubleTensor - empty
              fgradInput : DoubleTensor - empty
              finput : DoubleTensor - size: 75x15376
              gradBias : DoubleTensor - size: 16
              padH : 0
              weight : DoubleTensor - size: 16x75
              bias : DoubleTensor - size: 16
              gradWeight : DoubleTensor - size: 16x75
              padW : 0
              nInputPlane : 3
              kW : 5
              kH : 5
            }
        }
      output : DoubleTensor - size: 16x124x124
    }
}
```

File: training/train.lua (hunk 2 above)
Comment (melgor) [inline training/train.lua]: Here was the main issue of adding: multiGPU, optnet and clearState at once. 
The idea was taken from other [repo](https://github.com/facebook/fb.resnet.torch/pull/66) with explanation:
1. optimize memory with optnet, train
2. remove optnet optimizations
3. call clearState, save
go to step 1.

So optimization are removed and added every epoch, exactly is added [here](https://github.com/cmusatyalab/openface/pull/154/files#diff-dd628866be14580651da3058c63fb66eR39)

File: training/util.lua (hunk 3 above)
Comment (melgor) [inline training/util.lua]: I forgot about that functions. It was taken from fb.resnet repo. I can add second version of SpatialConvolutional but for your information they use exactly the [same algorithm](https://github.com/torch/nn/issues/159). So there is no reason to have both naming in provided models. What do you think?

File: training/model.lua (hunk 1 above)
Comment (melgor) [inline training/model.lua]: This would be hardest part:) I will try to bring back the "float" option. Maybe now when 3 modules works(multi-GPU, optnet and clearState), now just one stayed (float).
